staging/src/k8s.io/apimachinery/pkg/util/cache/lruexpirecache_test.go

```diff
@@ -66,3 +73,53 @@ func TestLRUOverflow(t *testing.T) {
 	expectEntry(t, c, "elem4", "4")
 	expectEntry(t, c, "elem5", "5")
 }
+
+func TestAddBringsToFront(t *testing.T) {
+	c := NewLRUExpireCache(4)
+	c.Add("elem1", "1", 10*time.Hour)
+	c.Add("elem2", "2", 10*time.Hour)
+	c.Add("elem3", "3", 10*time.Hour)
+	c.Add("elem4", "4", 10*time.Hour)
+
+	c.Add("elem1", "1-new", 10*time.Hour)
+
+	c.Add("elem5", "5", 10*time.Hour)
+
+	expectNotEntry(t, c, "elem2")
+	expectEntry(t, c, "elem1", "1-new")
+	expectEntry(t, c, "elem3", "3")
+	expectEntry(t, c, "elem4", "4")
+	expectEntry(t, c, "elem5", "5")
+}
+
+func TestGetBringsToFront(t *testing.T) {
+	c := NewLRUExpireCache(4)
+	c.Add("elem1", "1", 10*time.Hour)
+	c.Add("elem2", "2", 10*time.Hour)
+	c.Add("elem3", "3", 10*time.Hour)
+	c.Add("elem4", "4", 10*time.Hour)
+
+	c.Get("elem1")
+
+	c.Add("elem5", "5", 10*time.Hour)
+
+	expectNotEntry(t, c, "elem2")
+	expectEntry(t, c, "elem1", "1")
+	expectEntry(t, c, "elem3", "3")
+	expectEntry(t, c, "elem4", "4")
+	expectEntry(t, c, "elem5", "5")
+}
+
+func TestLRUKeys(t *testing.T) {
+	c := NewLRUExpireCache(4)
+	c.Add("elem1", "1", 10*time.Hour)
+	c.Add("elem2", "2", 10*time.Hour)
+	c.Add("elem3", "3", 10*time.Hour)
+	c.Add("elem4", "4", 10*time.Hour)
+
+	gotKeys := c.Keys()
+	wantKeys := []interface{}{"elem1", "elem2", "elem3", "elem4"}
```

Comment:
nit: could you also verify the keys in the other tests (make the logic below re-usable, e.g. `func assertKeys(t *testing.T, gotKeys []interface{}, wantKeys ...interface{})`). This would exercise the `Keys()` function in more places, to make sure the cache has correct size and ordering.

Reply:
I am implementing this, but it has turned up a slight problem.  Right now, the documentation for `LRUExpireCache.Keys()` says that it is allowed to return keys that are currently expired.  This makes sense in general because a caller won't have a guarantee that a particular key hasn't expired by the time they act on it from the result of `Keys()`.

However, `LRUExpireCache.Get()` proactively removes expired keys from the expiry list.

I was worried that this would make the tests a bit fragile, so I slightly changed the contract of `Keys()` so that it will not return presently-expired keys.